### ISSUE TITLE
continue doesn't work when debugging with breakpoints

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1928]
I have a function that I want to debug, so I add `error 0` as a rudimentary breakpoint:
```m2
f = x -> (
    print x;
    error 0;
    print x;
x)
```
Entering `help` in the debugger says:
```
           continue            -- leave the debugger, continuing execution
                               -- with current expression
```
Unless I'm misinterpreting this, it means if I enter `continue`, that means continue until the next breakpoint. But this has never worked for me:
```m2
i1 : load "test.m2"

i2 : errorDepth=1

o2 = 1

i3 : f(12)
12
test.m2:3:5:(3):[1]: error: 0
test.m2:3:5:(3):[1]: --entering debugger (type help to see debugger commands)
test.m2:3:5-3:11: --source code:
    error 0;

ii4 : continue
test.m2:3:5:(3):[1]: error: 0
test.m2:3:5:(3):[1]: --entering debugger (type help to see debugger commands)
test.m2:3:5-3:11: --source code:
    error 0;

ii5 : continue
test.m2:3:5:(3):[1]: error: 0
test.m2:3:5:(3):[1]: --entering debugger (type help to see debugger commands)
test.m2:3:5-3:11: --source code:
    error 0;

ii6 : return
test.m2:4:5:(3):[1]: --stepping limit reached
test.m2:4:5:(3):[1]: --entering debugger (type help to see debugger commands)
test.m2:4:5-4:11: --source code:
    print x;

ii7 : return
test.m2:5:1:(3):[1]: --stepping limit reached
test.m2:5:1:(3):[1]: --entering debugger (type help to see debugger commands)
test.m2:5:1-5:1: --source code:
x)

ii8 : return
stdio:3:1:(3): --done stepping, returning to top level

i9 : oo

o9 = 1
```
Note that if either `continue` or `return` had worked as I described, I should have got `12` as the output. Am I misunderstanding what `continue` does? At the very least, the documentation begins with "leave the debugger", and that didn't happen, so it should be clarified and a facility for what I described above should be added.

**Comment by DanGrayson** [comment]
`continue` continues by re-executing the current expression.  The idea is that you might have changed some values of some variables that allow the current expression to succeed.  You can use `return` to return something else as the value of `error 0`.

```m2

i8 : f 12
12
/tmp/foo.m2:3:5:(3):[1]: error: 0
/tmp/foo.m2:3:5:(3):[1]: --entering debugger (type help to see debugger commands)
/tmp/foo.m2:3:5-3:11: --source code:
    error 0;

ii9 : return
/tmp/foo.m2:4:5:(3):[1]: --stepping limit reached
/tmp/foo.m2:4:5:(3):[1]: --entering debugger (type help to see debugger commands)
/tmp/foo.m2:4:5-4:11: --source code:
    print x;

ii10 : continue
12

o10 = 12
```

**Comment by mahrud** [comment]
I think there's an off-by-one difference between M2 and other debuggers I've used. I would expect `return` to end that function by return that value and move up in the stack. Exactly as if I had changed the function and written `return` there. Does that make sense?

**Comment by mahrud** [comment]
Also, I haven't seen this behavior in other debuggers:
> `continue` continues by re-executing the current expression.

Maybe `continue` should be renamed to `redo` or `rerun`.

**Comment by DanGrayson** [comment]
Yes, "continue" doesn't convey the idea of "redoing", but similarly, "redo" doesn't convey the idea of "continuing".

Maybe "return" could be renamed to "provide", because it provides a value for the current expression.
```m2

i12 : code f

o12 = /tmp/foo.m2:1:7-1:21: --source code:
      f = x -> 10 + 1/0 + 100

i13 : f ()
/tmp/foo.m2:1:16:(3):[1]: error: division by zero
/tmp/foo.m2:1:16:(3):[1]: --entering debugger (type help to see debugger commands)
/tmp/foo.m2:1:15-1:17: --source code:
f = x -> 10 + 1/0 + 100

ii14 : return 1000
/tmp/foo.m2:1:21:(3):[1]: --stepping limit reached
/tmp/foo.m2:1:21:(3):[1]: --entering debugger (type help to see debugger commands)
/tmp/foo.m2:1:21-1:21: --source code:
f = x -> 10 + 1/0 + 100

ii15 : continue

o15 = 1110
```

**Comment by DanGrayson** [comment]
Hmm, in the example above, I think "continue" did not "redo" anything!

**Comment by mahrud** [comment]
Exactly! I think continue should never redo anything, but if you want such a command, a new one should be added.

On a c/c++ debugger, when a breakpoint is set on a line, execution stops before running that line, so when you step or continue, it is run for the first time. The same doesn't hold in M2 yet, but perhaps a breakpoint keyword would be useful.

**Comment by DanGrayson** [comment]
Here's how I set a breakpoint:
```m2
    error "debug me";
```

**Comment by mahrud** [comment]
Dan, I'm aware!! The point is that `continue` is not supposed to get stuck on a breakpoint, so it should be fixed. That's literally the title of the issue!

**Comment by mahrud** [comment]
Debugging would be much easier if this issue was resolved by either:
- change `continue` to not rerun the current expression again
- add a `breakpoint` keyword that `continue` will skip over

I just want to be able to repeatedly type continue and go forward. Typing `return <enter> continue <enter> ...` is too cumbersome.

**Comment by DanGrayson** [comment]
Sounds good, but for the first option, we'd still want a way to "redo" the current expression.

**Comment by mahrud** [comment]
Sure. Where are these keywords handled?

**Comment by DanGrayson** [comment]
Inside the function `handleError` in the file `d/evaluate.d`.

**Comment by d-torrance** [comment]
> **Written by Claude** (Claude Opus 5, via Claude Code), not by @d-torrance, whose account posted it -- please weigh it accordingly.

The second of the two options offered here exists now, and @mahrud added it himself:
`59ea23120c` introduced the `breakpoint` keyword, which shipped in #3431 and first appeared in
1.25.05. That was a pull request about top-level profiling, which is why nothing ever linked it
back to this issue. `interp.dd` handles the `continue` command from a breakpoint by evaluating
the code rather than re-entering the debugger, so on 1.26.06:

```m2
i2 : f 12
12
bp.m2:3:20:(3): entering debugger (enter 'help' to see commands)
bp.m2:3:15-3:36: --source code:
    breakpoint print "at breakpoint";

ii3 : continue
bp.m2:3:20:(3):[2]: unhandled continue command
at breakpoint
12

o3 = 12
```

That `12` is the output this issue said it should have got.

Leaving the issue open for three loose ends, all of them in the part of the report that asked
for the behaviour to be clarified rather than changed:

- `error 0` as a breakpoint is unchanged: `continue` still re-runs it and loops.
- The debugger help still reads `continue -- leave the debugger, continuing execution with
  current expression` (`m2/code.m2:261`), which is the wording called out above as not matching
  what happens.
- `continue` at a real breakpoint prints `unhandled continue command` before doing the right
  thing.

<sub>Triaged in a sweep of the open issues; source of truth: [`bug-triage/issues.tsv`](https://github.com/d-torrance/M2/blob/bug-triage/bug-triage/issues.tsv).</sub>
